In [ ]:
# dependencies
!pip install huggingface_hub
!pip install timm

from google.colab import drive
from huggingface_hub import notebook_login
from transformers import AutoModelForCausalLM, AutoTokenizer
import numpy as np
import re
import torch
import timm

drive.mount('/content/drive')

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 22.0 MB/s eta 0:00:00
  Using cached nvidia_cuda_nvrtc_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (23.7 MB)
  Using cached nvidia_cuda_runtime_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (823 kB)
  Using cached nvidia_cuda_cupti_cu12-12.1.105-py3-none-manylinux1_x86_64.whl (14.1 MB)
  Using cached nvidia_cudnn_cu12-8.9.2.26-py3-none-manylinux1_x86_64.whl (731.7 MB)
  Using cached nvidia_cublas_cu12-12.1.3.1-py3-none-manylinux1_x86_64.whl (410.6 MB)
  Using cached nvidia_cufft_cu12-11.0.2.54-py3-none-manylinux1_x86_64.whl (121.6 MB)
  Using cached nvidia_curand_cu12-10.3.2.106-py3-none-manylinux1_x86_64.whl (56.5 MB)
  Using cached nvidia_cusolver_cu12-11.4.5.107-py3-none-manylinux1_x86_64.whl (124.2 MB)
  Using cached nvidia_cusparse_cu12-12.1.0.106-py3-none-manylinux1_x86_64.whl (196.0 MB)
  Using cached nvidia_nccl_cu12-2.19.3-py3-none-manylinux1_x86_64.whl (166.0 MB)
  Using cached nvidia_nvtx_cu12-12.1.105-py3-none-manylin

In [ ]:
# load from hugging face
#notebook_login()

#tokenizer = AutoTokenizer.from_pretrained("mistralai/Mistral-7B-v0.1")
#model = AutoModelForCausalLM.from_pretrained("mistralai/Mistral-7B-v0.1")

# save the model !!
#path = '/content/drive/My Drive/MyModels/mistral/Mixtral-7B-v0.1'
#model.save_pretrained(path)

In [ ]:
# load the model and tokenizer from drive
from transformers import CanineForMultipleChoice
model_choice = 'canine'
#model_choice = 'mistral'
#model_choice = 'mistral_instruct'

tokenizer = None
model = None

if model_choice == 'mistral':
    path_model = '/content/drive/My Drive/C5470prj/Mistral-7B-v0.1' #mistral simple
if model_choice == 'mistral_instruct':
    path_model = '/content/drive/My Drive/C5470prj/Mistral-7B-Instruct-v0.2' #mistral + instruct prompting
if model_choice == 'canine':
    path_model = "google/canine-s"

tokenizer = AutoTokenizer.from_pretrained(path_model)
model = AutoModelForCausalLM.from_pretrained(path_model)

Loading checkpoint shards:   0%|          | 0/3 [00:00<?, ?it/s]

In [ ]:
# for Mistral testing only
if 'Mistral' in model_choice:
    text = '''[INST] ### Instructions:
Below is an instruction that describes a
multiple choice task. Answer the following
multiple choice question by giving the
most appropriate response. Answer should
be one among options provided after the
question. Select the most suitable answer
while making the necessary assumptions.
Give only answer and a short explanation
of two or three sentences. Nothing else.

### Input:
Question: {A man shaves everyday, yet keeps his beard long.}
{a. He is a barber.}
{b. He wants to maintain his appearance.}
{c. He wants his girlfriend to buy him a razor.}
{d. None of above.}

### Output Format:
The correct answer is: {label} {explanation}
Change {label} to one of a, b, c or d in your output.
Change {explanation} to a short explanation.
Do not output anything else after it.  [/INST]

The correct answer is: '''

    inputs = tokenizer(text, return_tensors="pt")
# using Canine tokenizer:
#inputs = tokenizer(text, padding="longest", truncation=True, return_tensors="pt", return_token_type_ids=False)
# do this for different tokenizers!
#model.resize_token_embeddings(len(tokenizer))
    outputs = model.generate(**inputs, max_new_tokens=(50))

    print(tokenizer.decode(outputs[0], skip_special_tokens=True))

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


[INST] ### Instructions:
Below is an instruction that describes a
multiple choice task. Answer the following
multiple choice question by giving the
most appropriate response. Answer should
be one among options provided after the
question. Select the most suitable answer
while making the necessary assumptions.
Give only answer and a short explanation
of two or three sentences. Nothing else.

### Input:
Question: {A man shaves everyday, yet keeps his beard long.}
{a. He is a barber.}
{b. He wants to maintain his appearance.}
{c. He wants his girlfriend to buy him a razor.}
{d. None of above.}

Hint: The man can shave other people's beard.

### Output Format:
The correct answer is: {label} {explanation}
Change {label} to one of a, b, c or d in your output.
Change {explanation} to a short explanation.
Do not output anything else after it.  [/INST]

The correct answer is:  a. He is a barber.
Explanation: The man shaves everyday but keeps his own beard long, which is consistent with being a 

In [ ]:
import random
# Load the data
SP_train = np.load('/content/drive/My Drive/C5470prj/BrainTeaser/data/SP-train.npy', allow_pickle=True)
WP_train = np.load('/content/drive/My Drive/C5470prj/BrainTeaser/data/WP-train.npy', allow_pickle=True)

SP_train = SP_train[:49]  # if recover from crashes
WP_train = WP_train[:49]

# Function to format the input text according to the model's required format
def format_question(data):
    id = data['id']
    question = data['question']
    print("processing data id: " + id + " length: " + str(len(question)))

    choices = data['choice_list']
    choice_order = data['choice_order']
    formatted_choices = [f"{chr(97 + i)}. {choices[idx]}" for i, idx in enumerate(choice_order)]
    text = f"""###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: {question}
    {formatted_choices[0]}
    {formatted_choices[1]}
    {formatted_choices[2]}
    {formatted_choices[3]}

    ### Output Format:
    The correct answer is: {{label}}{{explanation}}
    Change {{label}} to one of a, b, c or d in your output.
    Change {{explanation}} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is: """
    return text, id

# stats
correct_answers = 0
question_count = 0
SP_count = 0
SP_correct = 0
WP_count = 0
WP_correct = 0
wp = False
sp = False

# do this for different tokenizers!
#model.resize_token_embeddings(len(tokenizer))

for data in np.concatenate((SP_train, WP_train)):
    wp = False
    sp = False

    formatted_text, id = format_question(data)
    if '_' in id: # exclude reconstructed questions during eval
        continue
    if "WP" in id:
        wp = True
    if "SP" in id:
        sp = True

    question_count += 1
    if wp:
      WP_count += 1
    if sp:
      SP_count += 1

    # using mistral tokenizer:
    inputs = tokenizer(formatted_text, return_tensors="pt")
    # using Canine tokenizer:
    #inputs = tokenizer(formatted_text, padding="longest", truncation=True, return_tensors="pt", return_token_type_ids=False)

    # inference
    outputs = model.generate(**inputs, max_new_tokens=20) # 6 for base, 30 for instruct
    response = tokenizer.decode(outputs[0], skip_special_tokens=True)
    print(response)

    match_str = "The correct answer is:"
    predicted_answer = ''
    match = re.search(r'[^a-zA-Z]([a-d])[^a-zA-Z]', response[len(formatted_text):])
    if match:
        predicted_answer = match.group(1)
    else:
        predicted_answer = chr(97 + random.randint(0, 2))

    # Find the index of the correct answer in the choice order
    correct_label = ''
    # sanitize dataset
    if data['answer'] not in data['choice_list']:
        question_count -= 1
        if wp:
            WP_count -= 1
        if sp:
            SP_count -= 1
        continue

    correct_index = data['choice_list'].index(data['answer'])
    for i, ordered_i in enumerate(data['choice_order']):
        if ordered_i == correct_index:
            correct_label = chr(97 + i)

    if predicted_answer == correct_label:
        correct_answers += 1
        if wp:
            WP_correct += 1
        if sp:
            SP_correct += 1
    print("evaluated data id: " + id + ", predicted: " + predicted_answer + ", correct: " + correct_label + " total acc: " + str(correct_answers / question_count))
    if SP_count != 0:
        print(" SP acc: " + str(SP_correct / SP_count) + " correct: " + str(SP_correct) + "  count: "  + str(SP_count))
    if WP_count != 0:
        print(" WP acc: " + str(WP_correct / WP_count) + " correct: " + str(WP_correct) + "  count: "  + str(WP_count))

# Calculate accuracy
accuracy = correct_answers / question_count
print(f"Total Accuracy: {accuracy:.2f}")



Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


processing data id: SP-0 length: 139


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Mr. and Mrs. Mustard have six daughters and each daughter has one brother. But there are only 9 people in the family, how is that possible?
    a. Each daughter shares the same brother.
    b. Some daughters get married and have their own family.
    c. Some brothers were not loved by family and moved away.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The corr

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: A woman shoots her husband. Then she holds him underwater for over 5 minutes. Finally, she hangs him. But 5 minutes later, they both go out and enjoy a wonderful dinner together. How can this be?
    a. The woman gets a new partner.
    b. The woman was a photographer. She shot a picture of her husband, developed it, and hung it up to dry.
    c. The woman gets arrested for murder after dinner.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Samuel was out for a walk when it started to rain. He did not have an umbrella and he wasn't wearing a hat. His clothes were soaked, yet not a single hair on his head got wet. How could this happen?
    a. This man is bald.
    b. His hair is dyed.

    c. This man walk upside down to avoid rain.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer 

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Five apples are in a basket. How do you divide them among five girls so that each girl gets an apple, but one apple remains in the basket?
    a. One girl is polite and return her apple back.
    b. Give the fifth girl her apple in the basket.
    c. Two girls decide to share one apple.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  c {Tw

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Tom and his younger sister were fighting. Their mother was tired of the fighting and decided to punish them by making them stand on the same piece of newspaper in such a way that they couldn't touch each other. How did she accomplish this?
    a. Tom's mother slid a newspaper under a door, each sibling standing on each side.
    b. Tom and his younger sister stand on the newspaper back-to-back.
    c. The newspaper was folded multiple times and ended with a special shape.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
  

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: A horse is tied to a five-meter rope in front of an old saloon. Ten meters behind the horse is a bale of hay. Without breaking his rope, the horse is able to eat the hay whenever he chooses.
How is this possible?
    a. The walls of the saloon retract or collapse inwards, creating more space for the horse to reach the hay.
    b. The rope stretches proportionally, providing the extra length needed for the horse to reach the hay ten meters away.
    c. The rope is not tied to anything else.
    d. None of above.

    ### Output Format:
    The correct answer is: {labe

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: All of Mrs. Smith pets are dogs except one, and all her pets are cats except one. But Mrs.Smith only has 2 pets, how is that possible?
    a. Mrs.Smith has one hybrid pet that is a dog and a cat at the same time.
    b. Mrs. Smith has one cat and one dog.
    c. Mrs.Smith has one additional pet that is neither a dog or a cat.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: The 22nd and 24th presidents of the United States of America had the same parents, but were not brothers. How can this be possible?
    a. They were the same man.
    b. Their parents got divorced.
    c. The 24th president has better performance during tenure.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  b. Their parents got divorced.


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: There were eight girls going to school. They were sharing one umbrella. How come all the girls did not get wet?
    a. The umbrella that the girls were sharing was exceptionally large or had a wide canopy.
    b. The eight girls could have been sharing the umbrella at different times or taking turns using it.
    c. Because it was not raining.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: A black dog stands in the middle of an intersection in a town painted black. None of the street lights are working due to a power failure caused by a storm. A car with two broken headlights drives towards the dog but turns in time to avoid hitting him. How could the driver have seen the dog in time?
    a. It was during the daylight hours.
    b. The broken headlights can emit a small amount of light, even if it's dim. 
    c. The driver could have relied on their sense of hearing to detect the presence of the dog.
    d. None of above.

    ### Output Format:
    Th

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: On a fine sunny day a ship was in the harbor. All of a sudden the ship began to sink. There was no storm and nothing wrong with the ship yet it sank right in front of the spectators eyes. What caused the ship to sink?
    a. It was a Submarine.
    b. There are too many spectators in the harbor.
    c. There are too many fish around the ship.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything 

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: A man who lived on the top floor of a twenty story building had to go up and down daily for work, and of course, for food and the other necessities. On most days he could only ride the elevator to the fifteenth story, and he would have to walk the rest of the way. When it rained, however, he could ride all the way up to the twentieth story. Why?
    a. The man wasn't tall enough to reach the button for the 20th floor. He could only reach the 15th story button on the elevator. When it rained, he brought his umbrella, and used it to press the 20th story button on the e

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Even if they are starving, natives living in the Arctic will never eat a penguin's egg. Why not?
    a. Penguins only live in Antarctica.
    b. Penguins' eggs are toxic.
    c. Penguins and their eggs are not part of their traditional or cultural diet.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  c {Explanation: Natives living in the A

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: There once was an evil wizard. He took 3 woman from their homes and turned them into rose bushes that looked exactly alike. He put them in his garden. One of the woman had a husband and children and begged the wizard to let her see them. He agreed. At night, he brought the woman to her house. In the morning he came and took her home. One day the husband decided to go rescue her. So he stuck into the wizard's garden. He looked and looked at the 3 identical rose bushes trying to figure out which could be his wife. Suddenly, he knew the answer and he took his wife home.

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: There are 3 apples for 2 sons and 2 fathers to eat. Each of them get their own apple. How is it numerically possible?
    a. They are one son, one father and one grandfather.
    b. Two sons shared the same apple.
    c. One father gave his apple to his son.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  c. One father gave his apple to hi

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: A man moved into a new house and decided he was tired after the journey and turned off all the lights and went to bed. The next morning he killed 365 people, he does not sleep walk he slept like a baby. How could this be?
    a. He lived in a light house.
    b. The man is a world-renowned botanist who has just moved into a new house with a rare greenhouse containing 365 unique and endangered plant species.
    c. The man could be a calendar maker. He accidentally excluding February 29th on a leap year's calendar before sleep. This "killed" the day for 365 people.
  

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Joe and Sally were sitting in their family room one night. While Joe was watching T.V, his wife Sally was reading. All of a sudden the power went out and Joe decided to go to bed, but Sally kept on reading. With no use of artificial light, Sally kept on reading. How?
    a. Sally was blind, she was reading a book by Braille.
    b. Sally is smart that she pretended to reading the book.
    c. Because it was daytime.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
 

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: How do you spell COW in thirteen letters?
    a. COWCOWCOWCOWW
    b. SEE OH DEREFORD
    c. SEE O DOUBLE YOU.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  c. SEE O DOUBLE YOU.
    The letters in "COW" arranged to
evaluated data id: WP-0, predicted: c, correct: c total acc: 0.3333333333333333
 SP acc: 0.29411764705882354 correct: 5  cou

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: If eleven plus two equals one, what does nine plus five equal?
    a. Three.
    b. Four.
    c. Two.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    The given equation (11 + 2 = 1
evaluated data id: WP-1, predicted: d, correct: c total acc: 0.3157894736842105
 SP acc: 0.29411764705882354 correct: 5  count: 17
 WP acc

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: How can you take 2 from 5 and leave 4?
    a. Take the 5 and subtract 2 from it. This results in the number 3. Now, remove the vertical top part of the digit "3" and you are left with the digit "4." 
    b. FIVE. Remove the 2 letters F and E from five and you have in which is the Roman numeral for four.
    c. Imagine you have five fingers on your hand. If you take two fingers (your index and middle fingers) and hold them down, you are left with four fingers remaining upright. So, by taking 2 fingers from the set of 5 fingers, you are left with 4 fingers.
    d. None

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Tomorrow I am surely here, yesterday I am found as well. Today I am gone. Who am I?
    a. The Letter R.
    b. The Letter T.
    c. The Letter E.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    The answer is not a letter. The riddle describes the
evaluated data id: WP-4, predicted: d, correct: a total acc: 0.28571428

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What can you find in the center of gravity?
    a. The letter V.
    b. The letter G.
    c. The letter Y.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    The center of gravity is not associated with any specific letter.
evaluated data id: WP-5, predicted: d, correct: a total acc: 0.2727272727272727
 SP acc: 0.2941176

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What is in seasons, seconds, centuries and minutes but not in decades, years or days?
    a. The letter N.
    b. The letter D.
    c. The letter A.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    The given options do not match anything described in the question.
evaluated data id: WP-6, predicted: d, correct: a total

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: How many seconds are there in one year?
    a. 12 of them: January 2nd, February 2nd, March 2nd, April 2nd, May 2nd, June 2nd, July 2nd, August 2nd, September 2nd, October 2nd, November 2nd, December 2nd.
    b. One thousands.
    c. Six.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    Explanation: There are no "secon

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What arithmetic symbol can we place between 2 and 3 to make a number greater than 2 but less than 3?
    a. A division sign.
    b. A plus sign.
    c. A decimal point.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  b. A plus sign.
    Explanation: Adding a plus sign makes the number
evaluated data id: WP-8, predicted: b, correct: c total

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Do you know anything that has four eyes?
    a. Mississippi.
    b. Florida.
    c. California.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    Nothing in the given options has four eyes.
evaluated data id: WP-9, predicted: d, correct: a total acc: 0.23076923076923078
 SP acc: 0.29411764705882354 correct: 5  count: 17

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: How can you spell "enemy" in three letters?
    a. F O E.
    b. E N M.
    c. N M Y.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    The three-letter word "enemy" cannot be sp
evaluated data id: WP-10, predicted: d, correct: a total acc: 0.2222222222222222
 SP acc: 0.29411764705882354 correct: 5  count: 17
 WP acc: 0

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Ben walked into a hardware store and asked the price of some items. The salesman said: One costs $1, Eight costs $1, Seventeen cost $2, One hundred four costs $3 and One thousand seventy two costs $4. What was Ben buying?
    a. Ben was buying home address numbers and they cost $1 per digit.
    b. Ben is buying English word cards. The price is determined by the number of vowels in the spelled-out number.
    c. Ben is buying English word cards. The price is determined by the number of letters in the spelled-out number.
    d. None of above.

    ### Output Format:
 

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What liquid is represented by 'hijklmno'?
    a. Milk. "H L M N O" is similar as the nursery rhyme "High low，maids in a row", which commonly associated with milking cows. 
    b. Oil, if we rearrange the letters, they form the word "oil". 
    c. Water. Hijklmno is "H to O" in the alphabet. Water is H20.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: How can you say rabbit without using the letter R?
    a. Bunny.
    b. Abbit.
    c. White-animal with long things on head.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is: a. Bunny.
    Explanation: The word "rabbit" can be referred
evaluated data id: WP-13, predicted: a, correct: a total acc: 0.26666666666666666
 SP acc: 0.294117647058823

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: Which word in the English language becomes shorter when it is lengthened?
    a. Short.
    b. Truncation.
    c. Abbreviation.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  c. Abbreviation.
    An abbreviation is a short form of a
evaluated data id: WP-14, predicted: c, correct: a total acc: 0.25806451612903225
 SP acc: 0.29411764705882

Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What dress does everyone have, but no one wears?
    a. Address.
    b. Suit.
    c. Wedding dress.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  d. None of above.
    There is no dress that everyone has but no one wears
evaluated data id: WP-16, predicted: d, correct: a total acc: 0.25
 SP acc: 0.29411764705882354 correct: 5  count: 17


Setting `pad_token_id` to `eos_token_id`:2 for open-end generation.


###[INST]  Instructions:
    Below is an instruction that describes a multiple choice task. Answer the following multiple choice question by giving the most appropriate response. Answer should be one among options provided after the question. Select the most suitable answer while making the necessary assumptions. Give only answer and a short explanation of two or three sentences. Nothing else.

    ### Input:
    Question: What letters of the alphabet come too late for supper?
    a. Those that come after T.
    b. Those that at the end of alphabet.
    c. Those that at the middle of alphabet.
    d. None of above.

    ### Output Format:
    The correct answer is: {label}{explanation}
    Change {label} to one of a, b, c or d in your output.
    Change {explanation} to a short explanation.
    Do not output anything else after it. [/INST]

    The correct answer is:  a. Those that come after T.
    Explanation: The question asks for letters that
evaluated data id: WP-17, predicted: a,

In [ ]:
prompt = "In Italy, pizza served in formal settings, such as at a restaurant, is presented unsliced."
choice0 = "It is eaten with a fork and a knife."
choice1 = "It is eaten while held in the hand."
labels = torch.tensor(0).unsqueeze(0)  # choice0 is correct (according to Wikipedia ;)), batch size 1

encoding = tokenizer([prompt, prompt], [choice0, choice1], return_tensors="pt", padding=True)
outputs = model(**{k: v.unsqueeze(0) for k, v in encoding.items()}, labels=labels)  # batch size is 1

# the linear classifier still needs to be trained
loss = outputs.loss
logits = outputs.logits

##REsults
evaluated data id: WP-18, predicted: c, correct: a total acc: 0.2647058823529412
 SP acc: 0.29411764705882354 correct: 5  count: 17
 WP acc: 0.23529411764705882 correct: 4  count: 17
Total Accuracy: 0.26


evaluated data id: SP-208, predicted: a, correct: a total acc: 0.40236686390532544
 SP acc: 0.40236686390532544 correct: 68  count: 169

evaluated data id: WP-163, predicted: b, correct: a total acc: 0.3969465648854962
 WP acc: 0.3969465648854962 correct: 52  count: 131
Total Accuracy: 0.40



